### DISCUSSION TITLE
Will you deprecate getInitialProps?

### DISCUSSION THREAD

**Discussion by Fonger** [vercel/next.js#11775] [Help] [answered]
Currently we intensely use `getInitialProps`, and we even call it manually with custom HOC in nested child.

All data must be request-time generated. However, we don't rely on the value of getInitialProps.
Instead, we use redux and dispatch actions there hence redux store has value during SSR.

```js
static async getInitialProps({ store, res }) {
  await store.dispatch(action) // both client or server works by changing the store tree.
  return {}
}
```

This won't work if we switch to getServerSideProps or getStaticProps, right?

We just upgraded to next.js 9.3 and everything still works fine, but I'm concerned that `getInitialProps` is going to be deprecated and I have to refactor the code again.

**Comment by dheeraj-p** [comment]
You can still get it working with `getServerSideProps`. The only thing is `getServerSideProps` is always called on server so you can't write client side code in it. If you are not writing any client side code in `getInitialProps` then you shouldn't have many problems.

**Reply by ingoclaro** [reply]
there is a big difference though: you need to export getServerSideProps from the file of the Page.
getInitialProps can be added by a HOC like this:

```
const withMyCommonStuff = require(....

const Page = () => ...

export default withMyCommonStuff(Page)
```

this isn't possible with getServerSideProps

**Reply by dheeraj-p** [reply]
You should be able to do this 
```
const withMyCommonStuff = require(.... // Your logic for getInitialProps/getServerSideProps
const Page = () => ...
export default Page;
expor async function getServerSideProps(ctx){
  withMyCommonStuff() 
  ..
  ..
  return someProps
}
```

**Reply by ingoclaro** [reply]
only for server side rendering though, if I have client side rendering behavior then I also need to put something inside the component itself, and make sure it is not run for the SSR scenario...

**Reply by dheeraj-p** [reply]
Yes. I agree. It's not a smooth transition from getInitialProps to getServerSideProps.

**Reply by Fonger** [reply]
> You can still get it working with `getServerSideProps`. The only thing is `getServerSideProps` is always called on server so you can't write client side code in it. If you are not writing any client side code in `getInitialProps` then you shouldn't have many problems.

I use redux in getInitialProps, it’s server side code, but I don’t rely on the return value. Instead, I mutate the redux store there. If it runs on server, it has no way to let clients know they should mutate the redux tree.

I can still do this with processing the return value of server props, and handle it in client to update redux store. However, in this case, all requests that I await in getServerSideProps() on server side is in single request. And if everything was from server side, I can’t cache our API endpoints, which are behind CloudFlare CDN. I really hope getInitialProps won’t be deprecated since it’s a big breaking change.

**Comment by timneutkens** [comment, marked as answer]
`getInitialProps` is not deprecated. We do recommend using the new data fetching methods for the majority of use cases.

**Reply by Fonger** [reply]
Thanks. I hope it won't be deprecated in future either as it will be a breaking change.

**Reply by timneutkens** [reply]
It's very easy for us to support it, so we're not planning to remove it at this point in time.
